src/herder/TxSetFrame.h

```diff
@@ -7,115 +7,88 @@
 #include "ledger/LedgerHashUtils.h"
 #include "overlay/StellarXDR.h"
 #include "transactions/TransactionFrame.h"
-#include "util/UnorderedMap.h"
 #include <deque>
 #include <functional>
-#include <optional>
 
 namespace stellar
 {
 class Application;
 
 class TxSetFrame;
-typedef std::shared_ptr<TxSetFrame> TxSetFramePtr;
-typedef std::shared_ptr<TxSetFrame const> TxSetFrameConstPtr;
+using TxSetFrameConstPtr = std::shared_ptr<TxSetFrame const>;
 
-class AbstractTxSetFrameForApply
+// A wrapper for a set of transactions that if valid, maintains the hash order
+class TxSetFrame
 {
   public:
-    virtual ~AbstractTxSetFrameForApply(){};
-
-    virtual int64_t getBaseFee(LedgerHeader const& lh) const = 0;
-
-    virtual Hash const& getContentsHash() = 0;
-
-    virtual Hash const& previousLedgerHash() const = 0;
-
-    virtual size_t sizeTx() const = 0;
-
-    virtual size_t sizeOp() const = 0;
-
-    virtual std::vector<TransactionFrameBasePtr> sortForApply() = 0;
-    virtual void toXDR(TransactionSet& set) = 0;
-};
-
-class TxSetFrame : public AbstractTxSetFrameForApply
-{
-    std::optional<Hash> mHash;
-
-    // mValid caches both the last app LCL that we checked
-    // vaidity for, and the result of that validity check.
-    std::optional<std::pair<Hash, bool>> mValid;
-
-    Hash mPreviousLedgerHash;
-
     using AccountTransactionQueue = std::deque<TransactionFrameBasePtr>;
+    using Transactions = std::vector<TransactionFrameBasePtr>;
 
-    bool checkOrTrim(Application& app,
-                     std::vector<TransactionFrameBasePtr>& trimmed,
-                     bool justCheck, uint64_t lowerBoundCloseTimeOffset,
-                     uint64_t upperBoundCloseTimeOffset);
-
-    UnorderedMap<AccountID, AccountTransactionQueue> buildAccountTxQueues();
-    friend struct SurgeCompare;
-
-  public:
-    std::vector<TransactionFrameBasePtr> mTransactions;
+    TxSetFrame(Hash const& previousLedgerHash,
+               Transactions const& transactions);
 
     TxSetFrame(Hash const& previousLedgerHash);
 
-    TxSetFrame(TxSetFrame const& other) = default;
-
     // make it from the wire
     TxSetFrame(Hash const& networkID, TransactionSet const& xdrSet);
 
+    TxSetFrame(TxSetFrame const& other) = default;
+
     virtual ~TxSetFrame(){};
 
     // returns the hash of this tx set
-    Hash const& getContentsHash() override;
+    Hash const&
+    getContentsHash() const
+    {
+        releaseAssert(mTxsInValidHashOrder);
```

Comment:
I'm decoupling xdr validation from `TxSetFrame` construction in my CAP-42 PR (as it doesn't make sense to even try building a `TxSetFrame` around some malformed message). Maybe it should happen here instead, so that we have an assert to ensure that we always build correct frames.